File: 803-cheapest-flights-within-k-stops/solution.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        k += 1
        self.src, self.dst = src, dst
        self.memo = [[-888]*(k+1) for _ in range(n)]
        self.indegree = dict()

        for f in flights:
            _from, _to, _price = f
            if _to not in self.indegree:
                self.indegree[_to] = [(_from, _price)]
            else:
                self.indegree[_to].append((_from, _price))
        
        def dp(s, k):
            if s == self.src: return 0
            if k == 0: return -1
            if self.memo[s][k] != -888: return self.memo[s][k]

            res = sys.maxsize
            if s in self.indegree:
                for _from, _price in self.indegree[s]:
                    sub_problem = dp(_from, k-1)
                    if sub_problem != -1:
                        res = min(res, sub_problem + _price)
            self.memo[s][k] = -1 if res == sys.maxsize else res
            return self.memo[s][k]
        return dp(dst, k)

File: 803-cheapest-flights-within-k-stops/solution.java
class Solution {
    HashMap<Integer, List<int[]>> indegree;
    int src, dst;
    // 备忘录
    int[][] memo;

    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int K) {
        // 将中转站个数转化成边的条数
        K++;
        this.src = src;
        this.dst = dst;
        // 初始化备忘录，全部填一个特殊值
        memo = new int[n][K + 1];
        for (int[] row : memo) {
            Arrays.fill(row, -888);
        }

        indegree = new HashMap<>();
        for (int[] f : flights) {
            int from = f[0];
            int to = f[1];
            int price = f[2];
            // 记录谁指向该节点，以及之间的权重
            indegree.putIfAbsent(to, new LinkedList<>());
            indegree.get(to).add(new int[]{from, price});
        }

        return dp(dst, K);
    }

    // 定义：从 src 出发，k 步之内到达 s 的最短路径权重
    int dp(int s, int k) {
        // base case
        if (s == src) {
            return 0;
        }
        if (k == 0) {
            return -1;
        }
        // 查备忘录，防止冗余计算
        if (memo[s][k] != -888) {
            return memo[s][k];
        }

        // 初始化为最大值，方便等会取最小值
        int res = Integer.MAX_VALUE;
        if (indegree.containsKey(s)) {
            // 当 s 有入度节点时，分解为子问题
            for (int[] v : indegree.get(s)) {
                int from = v[0];
                int price = v[1];
                // 从 src 到达相邻的入度节点所需的最短路径权重
                int subProblem = dp(from, k - 1);

                // 跳过无解的情况
                if (subProblem != -1) {
                    res = Math.min(res, subProblem + price);
                }
            }
        }
        // 存入备忘录
        memo[s][k] = res == Integer.MAX_VALUE ? -1 : res;
        return memo[s][k];
    }
}